src/coreclr/jit/objectalloc.cpp

```diff
@@ -1335,3 +1483,1734 @@ void ObjectAllocator::RewriteUses()
+    // toVisitTryEntry: subset of above that are try entries.
+    //
+    CompAllocator                alloc(comp->getAllocator(CMK_ObjectAllocator));
+    ArrayStack<BasicBlock*>      toVisit(alloc);
+    jitstd::vector<BasicBlock*>* visited         = new (alloc) jitstd::vector<BasicBlock*>(alloc);
+    jitstd::vector<BasicBlock*>* toVisitTryEntry = new (alloc) jitstd::vector<BasicBlock*>(alloc);
+
+    BitVecTraits traits(comp->compBasicBlockID, comp);
+    BitVec       visitedBlocks(BitVecOps::MakeEmpty(&traits));
+    toVisit.Push(allocBlock);
+
+    // todo -- some kind of runaway size limit
+    //
+    while (toVisit.Height() > 0)
+    {
+        BasicBlock* const visitBlock = toVisit.Pop();
+        if (!BitVecOps::TryAddElemD(&traits, visitedBlocks, visitBlock->bbID))
+        {
+            continue;
+        }
+
+        if (visitBlock != allocBlock)
+        {
+            visited->push_back(visitBlock);
+        }
+
+        // We expect this stretch of blocks to all be in the same EH region.
+        //
+        if (!BasicBlock::sameEHRegion(allocBlock, visitBlock))
+        {
+            JITDUMP("Unexpected: new EH region at " FMT_BB "\n", visitBlock->bbNum);
+            return false;
+        }
+
+        if (visitBlock == defBlock)
+        {
+            continue;
+        }
+
+        JITDUMP("walking through " FMT_BB "\n", visitBlock->bbNum);
+
+        for (BasicBlock* const succ : visitBlock->Succs())
+        {
+            if (BitVecOps::IsMember(&traits, visitedBlocks, succ->bbID))
+            {
+                continue;
+            }
+            toVisit.Push(succ);
+        }
+    }
+
+    JITDUMP("def block " FMT_BB " post-dominates allocation site " FMT_BB "\n", defBlock->bbNum, allocBlock->bbNum);
+
+    // -1 here since we won't need to clone the allocation site itself.
+    //
+    JITDUMP("allocation side cloning: %u blocks\n", visited->size() - 1);
+
+    // The allocationBlock should not dominate the defBlock.
+    // (if it does, optimization does not require cloning, as
+    // there should be only one reaching def...)
+    //
+    if (comp->m_domTree->Dominates(allocBlock, defBlock))
+    {
+        JITDUMP("Unexpected, alloc site " FMT_BB " dominates def block " FMT_BB "\n", allocBlock->bbNum,
+                defBlock->bbNum);
+
+        return false;
+    }
+
+    auto latestStmt = [](Statement* stmt1, Statement* stmt2) {
+        if (stmt1 == stmt2)
+        {
+            return stmt1;
+        }
+
+        Statement* cursor1 = stmt1->GetNextStmt();
+        Statement* cursor2 = stmt2->GetNextStmt();
+
+        while (true)
+        {
+            if ((cursor1 == stmt2) || (cursor2 == nullptr))
+            {
+                return stmt2;
+            }
+
+            if ((cursor2 == stmt1) || (cursor1 == nullptr))
+            {
+                return stmt1;
+            }
+
+            cursor1 = cursor1->GetNextStmt();
+            cursor2 = cursor2->GetNextStmt();
+        }
+    };
```

Comment:
Extract this and the other versions into a `Compiler` method?

Reply:
Did so.